File: backend (Spring Boot)/cinemax/src/main/java/com/project/cinemax/service/TicketService.java
package com.project.cinemax.service;

import java.util.Date;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import com.project.cinemax.controller.TicketController.RecivedObjectforTicket;
import com.project.cinemax.model.Cinema;
import com.project.cinemax.model.Ticket;
import com.project.cinemax.model.User;
import com.project.cinemax.repository.CinemaRepository;
import com.project.cinemax.repository.TicketRepository;
import com.project.cinemax.repository.UserRepository;
import com.project.cinemax.specification.TicketSpecification;

@Service
public class TicketService {
	
	@Autowired
	private CinemaRepository cinemaRepository;
	
	@Autowired
	private UserRepository userRepository;
	
	@Autowired
	private TicketRepository ticketRepository;
	
	
	@Scheduled(fixedRate = 1000 * 60 * 60 * 12)
	public void deleteOldTickets() {
		Date oneDaysBefore = new Date(System.currentTimeMillis()-24*60*60*1000);
		List<Ticket> tickets = ticketRepository.findByProjectionDateProjectionBefore(oneDaysBefore);
		ticketRepository.deleteAll(tickets);
		System.out.println("Old tickets deleted! (Scheduled)  " + new Date().toString());
	}
	
	public List<Ticket> getTicketsForUser(String username) {
		return ticketRepository.findByUserUserName(username);
	}
	
	public Page<Ticket> getTicketsPageForAdmin(String ownerFirstName,
			String ownerLastName, Date dateProjection, Long cinemaId, 
			String sortDateProjection, int page, int size) {
		Cinema cinema = cinemaRepository.findById(cinemaId).isPresent() ?
				cinemaRepository.findById(cinemaId).get() : null;
		Specification<Ticket> specification = TicketSpecification.getSpec(ownerFirstName, ownerLastName, cinema, dateProjection, sortDateProjection);
		Pageable paging = PageRequest.of(page, size);
		
		Page<Ticket> tickets = ticketRepository.findAll(specification, paging);
		return tickets;
	}
	
	public void deleteTicket(Long ticketId) {
		ticketRepository.deleteById(ticketId);
	}
	
	public List<Ticket> getTicketsByProjection(Long projectionId) {
		return ticketRepository.findByProjectionId(projectionId);
	}
	
	public Ticket bookTicket(RecivedObjectforTicket recivedObject) {
		User user = userRepository.findById(recivedObject.userName).get();
		Ticket t = new Ticket(recivedObject.price, user, recivedObject.selectedSeats, recivedObject.projection);
		Ticket t1 = ticketRepository.save(t);
		return t1;
	}

}

File: src/app/_services/ticket.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';

@Injectable({
  providedIn: 'root'
})
export class TicketService {

  host = 'http://localhost:8080';
  requestHeader = new HttpHeaders({ 'No-Auth': 'True' });

  constructor(private http: HttpClient) {
  }

  public getTicketsForUser(username: string) {
    return this.http.get(this.host + `/getTicketsForUser?username=${username}`);
  }

  getTicketsFiltered(currentPage: number, sort: string, ownerFirstName: string, 
    ownerLastName: string, dateProjection: any = 0, selectedCinemaId: number = 0) {
    if (dateProjection === null || dateProjection === 0) {
      return this.http.get(this.host + 
        `/ticketsForAdmin?page=${currentPage}&sortDateProjection=${sort}&ownerFirstName=${ownerFirstName}&ownerLastName=${ownerLastName}&cinemaId=${selectedCinemaId}`);
    } else {
      return this.http.get(this.host + 
        `/ticketsForAdmin?page=${currentPage}&sortDateProjection=${sort}&ownerFirstName=${ownerFirstName}&ownerLastName=${ownerLastName}&cinemaId=${selectedCinemaId}&dateProjection=${dateProjection}`);
    }
  }

  deleteTicket(ticketId: number) {
    return this.http.delete(this.host + `/deleteTicket?ticketId=${ticketId}`);
  }

  getTickets(projectionId: number) {
    return this.http.get(this.host + `/getTickets?projectionId=${projectionId}`);
  }

  bookTicket(price: any, userName: string, selectedSeats: any, projection: any) {
    let sendObject = {
      price: price,
      userName: userName,
      selectedSeats: selectedSeats,
      projection: projection
    }
    return this.http.post(this.host + `/bookTicket`, sendObject);
  }

}
